src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,286 @@
+import number_theory.modular
+import geometry.manifold.mfderiv
+import analysis.complex.upper_half_plane.functions_bounded_at_infty
+import number_theory.modular_forms.slash_actions
+
+/-!
+# Modular forms
+
+This file defines modular forms and proves some basic properties about them.
+
+We begin by defining the notion of weakly modular form from which we then we construct the vector
+space of modular forms, cusp forms and prove that the product of two modular forms is a modular
+form (of higher weight).
+-/
+
+universes u v
+
+open complex upper_half_plane
+
+open_locale topological_space manifold upper_half_plane
+
+noncomputable theory
+
+local notation `ℍ'` := (⟨_,
+ upper_half_plane.upper_half_plane_is_open⟩: topological_space.opens ℂ)
+
+local prefix `↑ₘ`:1024 := @coe _ (matrix (fin 2) (fin 2) _) _
+
+local notation `GL(` n `, ` R `)`⁺ := matrix.GL_pos (fin n) R
+
+local notation `SL(` n `, ` R `)` := matrix.special_linear_group (fin n) R
+
+variables (Γ : subgroup SL(2,ℤ)) (k : ℤ)
+
+namespace modular_forms
+
+local notation f `∣[`:73 k:0, A `]`  :72 := slash_action.map ℂ k A f
+
+lemma slash_action_eq_slash (k : ℤ) (A : Γ) (f : ℍ → ℂ) : f ∣[k, A] = slash k A f := rfl
+
+lemma slash_action_eq_slash' (k : ℤ) (A : SL(2, ℤ)) (f : ℍ → ℂ) : f ∣[k, A] = slash k A f := rfl
+
+/-- The space of functions that are weakly modular. -/
+def weakly_modular_submodule (k : ℤ) (Γ : subgroup SL(2,ℤ)) : submodule ℂ (ℍ → ℂ) :=
+{ carrier := { f : (ℍ → ℂ) | ∀ (γ : Γ), (f  ∣[k, γ]) = f },
+  zero_mem' := by {apply slash_action.mul_zero },
+  add_mem' := by { intros f g hf hg γ,
+    rw [slash_action.add_action k γ f g, hf γ, hg γ]},
+  smul_mem' := by { intros c f hf γ,
+    have : (c • f) ∣[k, γ] = c • (f ∣[k, γ]), by {apply slash_action.smul_action},
+    rw (hf γ) at this,
+    apply this,} }
+
+lemma weakly_modular_mem (k : ℤ) (Γ : subgroup SL(2,ℤ)) (f : ℍ → ℂ) :
+  f ∈ (weakly_modular_submodule k Γ) ↔ ∀ (γ : Γ), (f ∣[k, γ]) = f := iff.rfl
+
+lemma slash_mul (k1 k2 : ℤ) (A : GL(2, ℝ)⁺) (f g : ℍ → ℂ) :
+  (f * g) ∣[k1 + k2, A] = (((↑ₘ A).det) : ℝ) • (f ∣[k1, A]) * (g ∣[k2, A]) :=
+begin
+  ext1,
+  simp only [slash_action.map, slash, matrix.general_linear_group.coe_det_apply, subtype.val_eq_coe,
+   ←mul_assoc],
+  have h1 : ((((↑ₘ A).det) : ℝ)^(k1 + k2 - 1) : ℂ) =
+  (((↑ₘ A).det) : ℝ) * (((↑ₘ A).det) : ℝ)^(k1 - 1) * (((↑ₘ A).det) : ℝ)^(k2 - 1),
+    begin simp only [mul_assoc, matrix.general_linear_group.coe_det_apply, subtype.val_eq_coe,
+      coe_coe],
+    rw [←zpow_add₀, ←zpow_one_add₀],
+    ring_exp,
+    all_goals { norm_cast, apply (matrix.GL_pos.det_ne_zero A)},
+    end,
+  have h22 : (upper_half_plane.denom A x)^(-(k1+k2)) = (upper_half_plane.denom A x)^(-k1) *
+    (upper_half_plane.denom A x)^(-k2),
+  by { rw [int.neg_add, zpow_add₀], exact upper_half_plane.denom_ne_zero A x, },
+  rw [h1, h22],
+  simp only [slash, upper_half_plane.denom, pi.mul_apply, coe_coe, zpow_neg, algebra.smul_mul_assoc,
+    pi.smul_apply, real_smul],
```

Comment:
```suggestion
  simp only [slash_action.map, slash, matrix.general_linear_group.coe_det_apply, subtype.val_eq_coe,
    pi.mul_apply, pi.smul_apply, algebra.smul_mul_assoc, real_smul],
  set d : ℂ := ↑((↑ₘ A).det : ℝ),
  have h1 : d ^ (k1 + k2 - 1) = d * d ^ (k1 - 1) * d ^ (k2 - 1),
  { have : d ≠ 0,
    { dsimp [d],
      norm_cast,
      exact matrix.GL_pos.det_ne_zero A },
    rw [← zpow_one_add₀ this, ← zpow_add₀ this],
    ring_exp },
  have h22 : denom A x ^ (- (k1 + k2)) = denom A x ^ (- k1) * denom A x ^ (- k2),
  { rw [int.neg_add, zpow_add₀],
    exact upper_half_plane.denom_ne_zero A x, },
  rw [h1, h22],
```
Some tips for this kind of proof:
- reorganize to combine your simps
- introduce notation to make a purely algebraic step more readable